src/Registration/cReg/NiftiImageData.cpp

```diff
@@ -1626,6 +1626,37 @@ void NiftiImageData<dataType>::axpby(
         _data[i] = a * x._data[i] + b * y._data[i];
 }
 
+template<class dataType>
+void NiftiImageData<dataType>::xapyb(
+    const DataContainer& a_x, const void* ptr_a,
+    const DataContainer& a_y, const void* ptr_b)
+{
+	NiftiImageData<dataType>::axpby(ptr_a, a_x, ptr_b, a_y);
+}
+
+template<class dataType>
+void NiftiImageData<dataType>::xapyb(
+    const DataContainer& a_x, const DataContainer& a_a,
+    const DataContainer& a_y, const DataContainer& a_b)
+{
+    const NiftiImageData<dataType>& a = dynamic_cast<const NiftiImageData<dataType>&>(a_a);
+    const NiftiImageData<dataType>& b = dynamic_cast<const NiftiImageData<dataType>&>(a_b);
+    const NiftiImageData<dataType>& x = dynamic_cast<const NiftiImageData<dataType>&>(a_x);
+    const NiftiImageData<dataType>& y = dynamic_cast<const NiftiImageData<dataType>&>(a_y);
```

Comment:
could potentially throw unexpectedly if the type isn't ok. This is fine, but will confuse the user (certainly so a Python user).

should we cope with cases where the data-type isn't `NiftiImageData`? Might be quite hard. But if we don't, I think we should put the whole thing in a `try catch` block with some sensible error.

Also, in future, this can be shorter as
```c++
auto& a = dynamic_cast<const NiftiImageData<dataType>&>(a_a);
```

Reply:
At the python end `sapyb` for vectors is in a try/except so it will default to python if there's an issue.

It does seem a bit strange restricting a and b to the same types, but I don't know how data is organised so whether a numpy array (for example) is compatible with nifty or stir arrays.

Reply:
ok. Nevertheless, it's the C++ library that should give sensible errors as much as we can, not the "client" Python/MATLAB